text/0030-protocol-messages.md

```diff
@@ -0,0 +1,243 @@
++ Feature name: `protocol-messages`
++ Start date: 2020-04-27
++ RFC PR: [iotaledger/bee-rfcs#30](https://github.com/iotaledger/bee-rfcs/pull/30)
++ Bee issue: [iotaledger/bee#70](https://github.com/iotaledger/bee/issues/70)
+
+# Summary
+
+This RFC introduces the IOTA protocol messages.
+
+# Motivation
+
+To be able to take part in the IOTA networks, Bee nodes need to implement the exact same protocol presented in this RFC
+and currently being used by [IRI](https://github.com/iotaledger/iri) nodes and
+[HORNET](https://github.com/gohornet/hornet) nodes.
+
+# Detailed design
+
+## `Message` trait
+
+The `Message` trait is protocol agnostic and only provides serialization and deserialization to and from bytes buffers.
+It should not be used as is but rather be paired with a higher layer - like a type-length-value encoding - and as such
+does not provide any safety check on inputs/outputs.
+
+```rust
+/// A trait describing the behavior of a message.
+pub(crate) trait Message {
+    /// The unique identifier of the message within the protocol.
+    const ID: u8;
+
+    /// Returns the size range of the message as it can be compressed.
+    fn size_range() -> Range<usize>;
+
+    /// Deserializes a bytes buffer into a message.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be within the range returned by the `size_range` method.
+    fn from_bytes(bytes: &[u8]) -> Self;
+
+    /// Returns the size of the message.
+    fn size(&self) -> usize;
+
+    /// Serializes a message into a bytes buffer.
+    /// Panics if the provided buffer has an invalid size.
+    /// The size of the buffer should be equal to the one returned by the `size` method.
+    fn into_bytes(self, bytes: &mut [u8]);
+}
+```
+
+**Notes**:
+- `into_bytes` does not allocate a buffer because the following TLV protocol implies concatenating a header meaning
+  another allocation. Since this is a hot path, a slice of an already allocated buffer for both the header and payload
+  is expected; hence, limiting the amount of allocation to the bare minimum;
+- `from_bytes`/`into_bytes` panic if incorrectly used, only the following safe TLV module should directly use them;
+
+## Type-length-value protocol
+
+The [type-length-value](https://en.wikipedia.org/wiki/Type-length-value) module is a safe layer on top of the messages.
+It allows serialization/deserialization to/from bytes ready to be sent/received to/from a transport layer by prepending
+or reading a header containing the type and length of the payload.
+
+### Header
+
+```rust
+/// A header for the type-length-value encoding.
+pub(crate) struct Header {
```

Comment:
Let's discuss visibility modifiers. I am not sure that those belong into an RFC as those things change more often then we want. It has been the case during the prototype development that suddenly someone needed a certain trait/type to be public. In order to not having to keep the RFC in sync with the implementation all the time I would suggest not adding those modifiers to the RFC if they don't enhance the understanding of the proposed architecture.

Reply:
The point here is to make clear how the `Header` type is defined/represented internally. As a reader of this RFC I don't care so much about visibility of this type when trying to understand the design decision.

Reply:
What do you suggest ? Not putting a visibility modifier is actually choosing a visibility (public). I think I prefer going from private to public if needed rather than public to private which is a breaking change.

Reply:
Also, everything in this RFC is purposely private to the crate.

Reply:
that's wrong

Reply:
Yeah right sorry, following discussions at the same time and can't keep up. What is the real issue here ? Why do you think this is confusing ?

Reply:
By writing `pub(crate)` you also make a statement, that the implementation contains several modules which have to share this type.

Reply:
I am not saying that it's confusing. It's just not necessary to understand the RFC.

Reply:
And that if we should be careful about visibility statements of types in the RFC.

Reply:
What I wanted to say is that you have to make a choice. Not having it `pub(crate)` is a choice. I am making the choice that IMO is the most obvious because it currently reflects the actual code.

Reply:
How do you justify having a private type actually being used all over your crate ?

Reply:
IMHO the purpose of an RFC is not to restrict the implementor in such a way, that he can't change type visibility anymore, so leaving a way the modifier (although in a Rust program is also a statement) reduces some noise to get the idea of the design across. If you want to your type to keep its field private to the crate you can add `pub(crate)` to the field sure, but that doesn't give you any guarantess the implementor won't implement a `pub` function exposing this value anyway. You would have to explicitly mention that this RFC disallows such a public getter. Since this would all become too complicated I'ld suggest leaving the visibility modifiers away, and concentrate in an RFC on the type and trait modeling.

Reply:
But there is no "implementor", `Header` is not a trait, it's a struct. RFCs are about justifying the designs of the Bee team. Not a guide for someone to implement something. The `Header` should not leak outside of the crate. IMO removing it opens up a new kind of doubts and uncertainties. We are really bikeshedding here, there is literally no way this prevents someone from understanding the overall design.

Reply:
Unfortunately you missed the point I am trying to make. I am talking about the visibility modifier (see my initial comment).
Of course there is an implementor to an RFC, and in general the one implementing it is not identical with the person who wrote it. At least this is how I imagined the RFC process to work. I agree, that atm the situation is kind of twisted where we use RFCs after the fact to justify our implementation. But in the future those will turn into real proposals how to change/improve/extend the framework. And its a bit unfair tbh to say I made an argument this would prevent someone from understanding the overall design. Of course it does not. All I am arguing for is that visibility modifiers are not necessary to understand that design.

Reply:
If we remove it, there is the immediate question of why the implementation doesn't follow the specification. If we are fine answering that whenever it's being asked maybe it's alright. I just find it very weird and still don't understand by what it's being motivated. As I said, RFCs are not definitive, they can be updated if need be, so why not having them reflect the current state of the implementation ? Making an arbitrary choice assuming it will be easier to understand is highly subjective to me.

Reply:
well you have the same question if you change `pub(crate)` to `pub` and forget about updating the RFC, I don't see how what you do atm is any different from that.

You seem to want to make the RFCs so concrete, that we need to maintain two codebases. If we see RFCs as mere proposals that are **not** strict specifications we wouldn't have to do that.

Reply:
Remember that you also had doc comments in the RFC. If you had changed a tiny bit in such a comment, someone could ask you about it.

Reply:
No. I am just choosing the version that has the highest probably of staying like that. Header **is** `pub(crate)` and it probably will never change.

Anyway, I will remove.

Reply:
Such a high probability needs to be found for **any** proposed type in an RFC (not just for `Header`) if we go that route. But yea, I wanted to bring this up to hear your and other opinions on that. I am not someone who can't be convinced of the opposite :smile:

Reply:
Not only `Header` is `pub(crate)` in this RFC.

Reply:
"any proposed type in an(y) RFC" :trollface:

Reply:
I think we can close this. Thanks for the discussion, Thibault! :heart:

Reply:
I think I was going to do that for all my RFCs because it's the obvious thing to do for me

Reply:
yup, understandable. I was thinking about it for some time, until I realized that this would probably bite us at some point, because either we spend a lot of our time keeping the RFCs in sync with the code, or we'ld have "lying" RFCs. If we make absolutely clear, that RFCs only describe the *general* design decision, but aren't strict specifications, then I think the readers would still be fine with it, and we can focus on writing actual code.

Reply:
Anyways, as you said in the call we don't need to be fully consistent on the RFCs, so I apologize if that sounded a bit like it. I simply wanted to start a discussion about how to go about it because I just didn't know, but not force anyone to change what he's most comfortable with.

Reply:
Removed.